### PR TITLE
Fix for validate_presence_of for Rails6 secure_password problem

### PR DESCRIPTION
**Pull request by itsterry**
Addresses #1167 

Logic is that Rails 6 secure password attributes require an additional attribute called [attribute]_digest

So if we're in Rails 6 and the attribute has a matching [attribute]_digest attribute, it's a secure password

### CODE REVIEW

File: spec/support/unit/helpers/rails_versions.rb [hunk 1]
```diff
@@ -38,5 +38,13 @@ def rails_lt_5?
     def rails_5_x?
       rails_version =~ '~> 5.0'
     end
+
+    def rails_lt_6?
```
Comment (mcmire) [inline spec/support/unit/helpers/rails_versions.rb]: Good thinking, but what if we add these when we need them?

File: spec/unit/shoulda/matchers/active_model/validate_presence_of_matcher_spec.rb [hunk 2]
```diff
@@ -327,7 +327,7 @@ def model_creator
     end
   end
 
-  if rails_4_x?
+  if rails_4_x? || rails_5_x?
```
Comment (mcmire) [inline spec/unit/shoulda/matchers/active_model/validate_presence_of_matcher_spec.rb]: Good thinking. In order to merge this, we will also need tests for the `#{attribute}_digest` case. Unfortunately, in order for you to even run those tests, we will need to add Rails 6 to the Appraisals file. We can't do this quite yet until we have 4.0 released, so we will have to come back around to this.

File: spec/support/unit/helpers/rails_versions.rb (hunk 1 above)
Comment (mcmire) [reply spec/support/unit/helpers/rails_versions.rb]: Yeah, let's go ahead and leave them out for now, and we can add them when it comes time.

File: lib/shoulda/matchers/active_model/validate_presence_of_matcher.rb [hunk 3]
```diff
@@ -147,11 +147,21 @@ def simple_description
         private
 
         def secure_password_being_validated?
-          defined?(::ActiveModel::SecurePassword) &&
+          rails_lt_6_secure_password_being_validated? ||
+            rails_6_secure_password_being_validated?
+        end
+
+        def rails_lt_6_secure_password_being_validated?
```
[the remark this answers is not in the document]
Comment (mcmire) [reply lib/shoulda/matchers/active_model/validate_presence_of_matcher.rb]: If you don't mind changing this, that would be appreciated!

File: spec/support/unit/helpers/rails_versions.rb (hunk 1 above)
Comment (itsterry) [reply spec/support/unit/helpers/rails_versions.rb]: Have removed

File: lib/shoulda/matchers/active_model/validate_presence_of_matcher.rb (hunk 3 above)
[the remark this answers is not in the document]
Comment (itsterry) [reply lib/shoulda/matchers/active_model/validate_presence_of_matcher.rb]: Changed per feedback

File: lib/shoulda/matchers/active_model/validate_presence_of_matcher.rb [hunk 4]
```diff
@@ -146,14 +146,28 @@ def simple_description
...
-        def secure_password_being_validated?
-          defined?(::ActiveModel::SecurePassword) &&
-            @subject.class.ancestors.include?(::ActiveModel::SecurePassword::InstanceMethodsOnActivation) &&
+        def should_double_check_attribute?
+          attribute_is_secure_password? ||
+            attribute_has_digest?
+        end
+
+        def attribute_is_secure_password?
+          defined?(
```
Comment (ricardotk002) [inline lib/shoulda/matchers/active_model/validate_presence_of_matcher.rb]: I was reviewing Rails 6's new beta and stumbled with this issue too. I came up with the following question: shouldn't these checks be more optimistic about the attribute's name and check if it is `password` first, before doing the subsequent checks for `ActiveModel::SecurePassword`?

File: lib/shoulda/matchers/active_model/validate_presence_of_matcher.rb [hunk 5]
```diff
@@ -146,14 +146,28 @@ def simple_description
...
+            ::ActiveModel::SecurePassword::InstanceMethodsOnActivation
```
Comment (ricardotk002) [inline lib/shoulda/matchers/active_model/validate_presence_of_matcher.rb]: Btw, this module was removed in https://github.com/rails/rails/pull/26764

**Comment by mcmire** [comment]
Hi @itsterry we realized that #1117 already covered this! Sorry for not seeing this earlier. This is fixed in the PR above ^

**Comment by itsterry** [comment]
Brilliant! Thanks @mcmire !

**Pull request closed without merging**